packages/block-library/src/page-list/index.php

```diff
@@ -0,0 +1,217 @@
+		return;
+	}
+	$markup = '';
+	foreach ( (array) $nested_pages as $page ) {
+		$css_class = 'wp-block-pages-list__item';
+		if ( isset( $page['children'] ) ) {
+			$css_class .= ' has-child';
+		}
+		$markup .= '<li class="' . $css_class . '">';
+		$markup .= '<a class="wp-block-pages-list__item__link" href="' . esc_url( $page['link'] ) . '">' . wp_kses(
+			$page['title'],
+			wp_kses_allowed_html( 'post' )
+		) . '</a>';
+		if ( isset( $page['children'] ) ) {
+			$markup .= '<span class="wp-block-page-list__submenu-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 24 24" transform="rotate(90)"><path d="M8 5v14l11-7z"></path><path d="M0 0h24v24H0z" fill="none"></path></svg></span>';
+			$markup .= '<ul class="submenu-container">' . render_nested_page_list( $page['children'] ) . '</ul>';
+		}
+		$markup .= '</li>';
+	}
+	return $markup;
+}
+
+/**
+ * Outputs nested array of pages
+ *
+ * @param array $current_level The level being iterated through.
+ * @param array $children The children grouped by parent post ID.
+ *
+ * @return array The nested array of pages.
+ */
+function nest_pages( $current_level, $children ) {
+	if ( empty( $current_level ) ) {
+		return;
+	}
+	foreach ( (array) $current_level as $key => $current ) {
+		if ( isset( $children[ $key ] ) ) {
+			$current_level[ $key ]['children'] = nest_pages( $children[ $key ], $children );
+		}
+	}
+	return $current_level;
+}
+
+	/**
+	 * Renders the `core/page-list` block on server.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @param array $content The saved content.
+	 * @param array $block The parsed block.
+	 *
+	 * @return string Returns the page list markup.
+	 */
+function render_block_core_page_list( $attributes, $content, $block ) {
+	static $block_id = 0;
+	$block_id++;
+
+	$all_pages = get_pages();
+
+	$top_level_pages = array();
+
+	$pages_with_children = array();
+
+	foreach ( (array) $all_pages as $page ) {
+		if ( $page->post_parent ) {
+			$pages_with_children[ $page->post_parent ][ $page->ID ] = array(
+				'title' => $page->post_title,
+				'link'  => get_permalink( $page->ID ),
+			);
+		} else {
+			$top_level_pages[ $page->ID ] = array(
+				'title' => $page->post_title,
+				'link'  => get_permalink( $page->ID ),
+			);
+
+		}
+	}
+
+	$nested_pages = nest_pages( $top_level_pages, $pages_with_children );
+
+	$wrapper_markup = '<ul %1$s>%2$s</ul>';
+
+	$items_markup = render_nested_page_list( $nested_pages );
+
+	$colors          = block_core_page_list_build_css_colors( $block->context );
+	$font_sizes      = block_core_page_list_build_css_font_sizes( $block->context );
+	$classes         = array_merge(
+		$colors['css_classes'],
+		$font_sizes['css_classes']
+	);
+	$style_attribute = ( $colors['inline_styles'] || $font_sizes['inline_styles'] );
+	$css_classes     = trim( implode( ' ', $classes ) );
+
+	if ( $block->context && $block->context['showSubmenuIcon'] ) {
+		$css_classes .= ' show-submenu-icons';
+	}
```

Comment:
I would move this to it's own setter like `block_core_page_add_context_classes( $context, $classes )`.

Reply:
Would that include just this `if` statement, or the whole classname setting logic? The colours and font-size classes set above are also derived from context. 

With the submenu icons in particular, I'm hoping we can remove this class soon, and just render the submenu icons whenever they are needed. The problem is we're server side rendering the block in the editor, and `ServerSideRender` doesn't recognise context. There's some more discussion on this problem in #28684 . Anyway, I'd leave this as is for now and revisit later pending the outcome of that discussion.